### PR TITLE
Fix writing COOSYS elements from votable to xml

### LINKED ISSUE #17348: VOTable writer does not write out COOSYS except in RESOURCEs.
### Description

In the VOTable Format Definition document the Version History (Section 9.1) suggests that `COOSYS` was deprecated in VOTable v1.2, but it remains defined in Section 3.4 without note of deprecation, and in the Element Hierarchy in section 7.1 without the deprecation symbol. The current `COOSYS` tests place all `COOSYS` elements in `RESOURCE` elements, and any `COOSYS` elements placed outside of a `RESOURCE` element would be dropped as written as `VOTableFile.to_xml()` replaces `self.coordinate_systems` with `self.groups` if the version is 1.2 or later.

### Expected behavior

Unclear, but perhaps `COOSYS` elements should be reformatted as Space-Time Coordinate references on write.

### How to Reproduce

This script reads in `test_in.xml` which has `COOSYS` children for `VOTABLE` and `RESOURCE`.

Then it prints the contents of the `COOSYS` under the `VOTABLE` and `RESOURCE` to show that it has not been lost on read.

Then it writes the VOTable to `test_out.xml`. Note that the `COOSYS` under the `VOTABLE` is missing.

```python
from astropy.io.votable import parse

votable = parse("test_in.xml")
print(f"{votable.coordinate_systems=}")
print(f"{votable.resources[0].coordinate_systems=}")
votable.to_xml("test_out.xml")
```
Output:
```
votable.coordinate_systems=[<COOSYS ID="coosys0" epoch="J2015.5" equinox="J2000" refposition="BARYCENTER" system="ICRS"/>]
votable.resources[0].coordinate_systems=[<COOSYS ID="coosys1" epoch="J2015.5" equinox="J2000" refposition="BARYCENTER" system="ICRS"/>, <COOSYS ID="coosys2" epoch="B1950.0" equinox="B1950" refposition="UNKNOWN" system="FK4"/>]
```
test_in.xml:
```xml
<?xml version="1.0" encoding="UTF-8"?>
<VOTABLE xmlns="http://www.ivoa.net/xml/VOTable/v1.3" version="1.5">
  <COOSYS ID="coosys0" system="ICRS" equinox="J2000" epoch="J2015.5"  refposition="BARYCENTER"/>
  <RESOURCE>
  <!-- "equinox", "epoch", "system", "refposition" -->
    <COOSYS ID="coosys1" system="ICRS" equinox="J2000" epoch="J2015.5"  refposition="BARYCENTER"/>
    <COOSYS ID="coosys2" system="FK4" equinox="B1950" epoch="B1950.0"  refposition="UNKNOWN"/>
    <TABLE name="sources">
      <FIELD name="RA" ID="ra" ucd="pos.eq.ra;meta.main" datatype="double" unit="deg" ref="my_coosys"/>
      <FIELD name="Dec" ID="dec" ucd="pos.eq.dec;meta.main" datatype="double" unit="deg" ref="my_coosys"/>
      <FIELD name="Name" ID="name" ucd="meta.id;meta.main" datatype="char" arraysize="8*"/>
      <DATA>
        <TABLEDATA>
          <TR>
            <TD>010.68</TD>
            <TD>+41.27</TD>
            <TD>N 224</TD>
          </TR>
        </TABLEDATA>
      </DATA>
    </TABLE>
  </RESOURCE>
</VOTABLE>
```
test_out.xml
```xml
<?xml version="1.0" encoding="utf-8"?>
<!-- Produced with astropy.io.votable version 7.0.0.dev1484+g721e7e6222
     http://www.astropy.org/ -->
<VOTABLE version="1.5" xmlns="http://www.ivoa.net/xml/VOTable/v1.3" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.ivoa.net/xml/VOTable/v1.3 http://www.ivoa.net/xml/VOTable/VOTable-1.5.xsd">
 <RESOURCE type="results">
  <COOSYS ID="coosys1" epoch="J2015.5" equinox="J2000" refposition="BARYCENTER" system="ICRS"/>
  <COOSYS ID="coosys2" epoch="B1950.0" equinox="B1950" refposition="UNKNOWN" system="FK4"/>
  <TABLE ID="sources" name="sources" nrows="1">
   <FIELD ID="ra" datatype="double" name="RA" ref="my_coosys" ucd="pos.eq.ra;meta.main" unit="deg"/>
   <FIELD ID="dec" datatype="double" name="Dec" ref="my_coosys" ucd="pos.eq.dec;meta.main" unit="deg"/>
   <FIELD ID="name" arraysize="8" datatype="char" name="Name" ucd="meta.id;meta.main"/>
   <DATA>
    <TABLEDATA>
     <TR>
      <TD>10.68</TD>
      <TD>41.27</TD>
      <TD>N 224</TD>
     </TR>
    </TABLEDATA>
   </DATA>
  </TABLE>
 </RESOURCE>
</VOTABLE>
```

### Versions

```python
import astropy
try:
    astropy.system_info()
except AttributeError:
    import platform; print(platform.platform())
    import sys; print("Python", sys.version)
    import astropy; print("astropy", astropy.__version__)
    import numpy; print("Numpy", numpy.__version__)
    import erfa; print("pyerfa", erfa.__version__)
    try:
        import scipy
        print("Scipy", scipy.__version__)
    except ImportError:
        print("Scipy not installed")
    try:
        import matplotlib
        print("Matplotlib", matplotlib.__version__)
    except ImportError:
        print("Matplotlib not installed")
```
```
platform
--------
platform.platform() = 'macOS-14.7-arm64-arm-64bit-Mach-O'
platform.version() = 'Darwin Kernel Version 23.6.0: Wed Jul 31 20:48:04 PDT 2024; root:xnu-10063.141.1.700.5~1/RELEASE_ARM64_T6030'
platform.python_version() = '3.13.0'

packages
--------
astropy              7.0.0.dev1484+g721e7e6222
numpy                2.1.3
scipy                1.14.1
matplotlib           3.9.2
pandas               2.2.3
pyerfa               2.0.1.4
```

### PR DESCRIPTION
**Pull request by d-giles**
### Description

This pull request is to address inconsistent handling of COOSYS elements in writing to xml. COOSYS elements were deprecated in v1.2 and undeprecated in v1.4. However, COOSYS elements were being written in RESOURCE elements regardless of version, and base level COOSYS elements were removed from VOTableFile for every version greater than 1.2. <del>This PR implements a version check and removes COOSYS from VOTableFile and RESOURCE elements on writing to xml if the version is 1.2 or 1.3.</del>
Upon further review we've decided that the original removal of the COOSYS from VOTableFile was a mistake, so this PR implements the expected behavior of preserving COOSYS on both VOTableFile and RESOURCE elements regardless of VOTable version.

Fixes #17348

### CODE REVIEW

File: docs/changes/io.votable/17356.bugfix.rst [hunk 1]
```diff
@@ -0,0 +1 @@
+Updated xml writer for VOTableFile and Resource elements to include or drop coordinate_systems depending on version.
```
Comment (pllim) [inline docs/changes/io.votable/17356.bugfix.rst:1]:
```suggestion
Updated XML writer for ``VOTableFile`` and ``Resource`` elements to include or drop ``coordinate_systems`` depending on VOTable version.
```

File: astropy/io/votable/tree.py [hunk 2]
```diff
@@ -3978,6 +3978,8 @@ def to_xml(self, w, **kwargs):
             ]
             if kwargs["version_1_2_or_later"]:
                 element_sets.append(self.groups)
+                if not kwargs["version_1_4_or_later"]:
+                    del element_sets[0]
```
Comment (bsipocz) [inline astropy/io/votable/tree.py:3982]: I would find a comment here about the deprecation and undeprecation helpful. You do that below already, but there are already a lot of repetition in these methods, so the comment can be duplicated, too.

**Review by bsipocz — APPROVED** [review]
Looks all good to me, except a minor comment about a comment.

**Comment by tomdonaldson** [comment]
The changes here involved version-specific behavior, so I think it would be nice if the tests checked that behavior for the relevant versions.

**Comment by tomdonaldson** [comment]
I'll admit I'm not sure about the utility of removing the coordsys in these cases but I'll give it some more thought.

File: astropy/io/votable/tree.py (hunk 2 above)
Comment (d-giles) [reply astropy/io/votable/tree.py:3983]: Added in [f086bed](https://github.com/astropy/astropy/pull/17356/commits/f086bed3378177b11b1bb302596e7949daaf67d7)

File: astropy/io/votable/tree.py [hunk 3]
```diff
@@ -4378,7 +4381,11 @@ def to_xml(
                     self.resources,
                 ]
                 if kwargs["version_1_2_or_later"]:
-                    element_sets[0] = self.groups
+                    element_sets.append(self.groups)
+                    if not kwargs["version_1_4_or_later"]:
+                        # COOSYS was deprecated in v1.2, and undeprecated in v1.4
+                        del element_sets[0]
```
Comment (tomdonaldson) [inline astropy/io/votable/tree.py:4387]: I see that this deletion of the `coordinate_systems` was just preserving the previous behavior for the v1.2 and v1.3 while allowing it to go through for >=v1.4.  Then the new deletion above for `RESOURCE` is intending to match this behavior.

This is a pretty quirky case!  After more thinking about this, I believe that the original replacement here of coordinate_systems with groups was a mistake.  In fact `COOSYS` never became illegal, it was just deprecated, so it was ok to keep quietly supporting it.  

It's interesting that the original deletion here for `VOTABLE` was not mirrored for `RESOURCE`, but it definitely left us in an inconsistent state, and with the bug you caught in #17348.  I think the better fix here would be to just do the `append()` of the groups, but not the deletion, then not do the deletion in `RESOURCE`.

I apologize that I didn't write this before you went to the trouble of writing the extra version-specific testing!

Comment (d-giles) [reply astropy/io/votable/tree.py]: Not a problem, the updated tests were pretty straightforward.

I've gone ahead and implemented the change to keep `COOSYS` regardless of version and updated the truth reference to match the expected behavior. I rebased the whole PR to remove the unnecessary commits.

**Pull request merged**